### DISCUSSION TITLE
Support for remote private keys

### DISCUSSION THREAD

**Discussion by wiktor-k** [openpgpjs/openpgpjs#1432] [Ideas]
Hi :wave: 

I work on [Sequoia PGP](https://sequoia-pgp.org/) and just recently I've [added support for accessing remote private keys](https://gitlab.com/sequoia-pgp/sequoia/-/merge_requests/1149) (like keys on OpenPGP Cards, TPMs, PKCS#11 etc.). I've made a [mini-protocol](https://gitlab.com/wiktor/pks) for a small set of cryptographic primitives (basically RSA 1.5 sign/decrypt and ECDH derivation, ECDSA/EdDSA signing) accessible over HTTP.

I was wondering what do you think about including support for accessing keys via that protocol to OpenPGP.js?

If you think it's a good idea and could provide a couple of pointers I'm volunteering to send a PR :)

**Comment by twiss** [comment]
Hey :wave: This is an interesting idea :blush: The main thing I'm worried about in the context of OpenPGP.js, is adding code that won't be used by many users (i.e. distribution size is a significant concern for us). Perhaps we could think about making this some kind of plugin / add-on module?

That being said, I don't know if exposing enough hooks for this to be able to make it a separate module makes sense, or is easy to do in the context of our usage of ES6 modules. And perhaps if it's not that much code, it doesn't matter that much. But I think we should investigate this a bit.

Regardless of the above, there may be other benefits to having a concept of keys being "stored elsewhere", even for Web Crypto, for example: right now, we import keys into Web Crypto on every usage, and then throw them away again, but it would be cool to defer storing and using the key to (some code that invokes) Web Crypto entirely.

**Reply by wiktor-k** [reply]
> Regardless of the above, there may be other benefits to having a concept of keys being "stored elsewhere", even for Web Crypto, for example: right now, we import keys into Web Crypto on every usage, and then throw them away again, but it would be cool to defer storing and using the key to (some code that invokes) Web Crypto entirely.

Indeed. Web Crypto keys can be created with non exportable attribute (the key handle can be stored in IndexedDB) and this makes some interesting use cases possible e.g. generating one encryption subkey per browser/device. Device key management would be more transparent and you wouldn't have to store the encrypted subkey on the server side.

As for the private key protocol I've made it purposefully as simple as possible (comparable to WKD) but definitely if we had hooks or something like that to minimize the code size it's a good idea.

Maybe I can browse the code and propose something and then we can discuss that? (Unless you've got some specific approach in mind already)

**Reply by twiss** [reply]
Sounds good, go for it :+1: :blush:
